src/cmap/conn/wire/message.rs

```diff
@@ -75,33 +75,41 @@ impl Message {
     }
 
     /// Reads bytes from `reader` and deserializes them into a Message.
-    pub(crate) async fn read_from(reader: &mut AsyncStream) -> Result<Self> {
-        let header = Header::read_from(reader).await?;
-        let mut length_remaining = header.length - Header::LENGTH as i32;
-
+    pub(crate) async fn read_from(
+        reader: &mut AsyncStream,
+        partial_message_state: &mut PartialMessageState,
+    ) -> Result<Self> {
+        let length = reader.read_i32().await?;
+        partial_message_state.bytes_remaining = length as usize - Header::LENGTH;
```

Comment:
we're setting the bytes remaining before we actually performed the reads, and those reads may never occur if the future stops getting polled. I think we need to pass the partial message state down through `Head::read_from` as well, and make sure to update the bytes remaining with each read that we do.